R/methods-epi_archive.R

```diff
@@ -1000,3 +1000,79 @@ epix_truncate_versions_after.epi_archive <- function(x, max_version) {
   return((x$clone()$truncate_versions_after(max_version)))
   # ^ second set of parens drops invisibility
 }
+
+#' Print and summary functions for an `epi_archive` object.
+#'
+#' @param x The `epi_archive` object.
+#' @param class Boolean; whether to print the class label header
+#' @param methods Boolean; whether to print all available methods of
+#'   the archive
+#' @param ... Additional arguments passed to methods.
+#'
+#' @method print epi_archive
+#' @export
+print.epi_archive <- function(x, class = TRUE, methods = TRUE, ...) {
+  # --- Copied from R6 ---
+  if (class) cat("An `epi_archive` object, with metadata:\n")
+  cat(sprintf("* %-9s = %s\n", "geo_type", x$geo_type))
+  cat(sprintf("* %-9s = %s\n", "time_type", x$time_type))
+  if (!is.null(x$additional_metadata)) {
+    sapply(x$additional_metadata, function(m) {
+      cat(sprintf("* %-9s = %s\n", names(m), m))
+    })
+  }
+  cat("----------\n")
+  if (length(x$DT$time_value) == 0 || all(is.na(x$DT$time_value))) {
+    min_time <- max_time <- NA
+  } else {
+    min_time <- Min(x$DT$time_value)
+    max_time <- Max(x$DT$time_value)
+  }
+  cat(sprintf("* %-14s = %s\n", "min time value", min_time))
+  cat(sprintf("* %-14s = %s\n", "max time value", max_time))
+  cat(sprintf(
+    "* %-14s = %s\n", "first version with update",
+    min(x$DT$version)
+  ))
+  cat(sprintf(
+    "* %-14s = %s\n", "last version with update",
+    max(x$DT$version)
+  ))
+  if (is.na(x$clobberable_versions_start)) {
+    cat("* No clobberable versions\n")
+  } else {
+    cat(sprintf(
+      "* %-14s = %s\n", "clobberable versions start",
+      x$clobberable_versions_start
+    ))
+  }
+  cat(sprintf(
+    "* %-14s = %s\n", "versions end",
+    x$versions_end
+  ))
+  cat("----------\n")
+  cat(sprintf(
+    "Data archive (stored in DT field): %i x %i\n",
+    nrow(x$DT), ncol(x$DT)
+  ))
+  cat(sprintf("Columns in DT: %s\n", paste(ifelse(length(
+    colnames(x$DT)
+  ) <= 4, paste(colnames(x$DT), collapse = ", "),
+  paste(
+    paste(colnames(x$DT)[1:4], collapse = ", "), "and",
+    length(colnames(x$DT)[5:length(colnames(x$DT))]), "more columns"
+  )
+  ))))
+  if (methods) {
+    cat("----------\n")
+    writeLines(wrap_varnames(
+      initial = "Public R6 methods: ",
+      names(epi_archive$public_methods)
+    ))
+  }
+
+  # Error in NextMethod() : generic function not specified
+  # NextMethod()
+
+  x$DT %>% tibble %>% print
```

Comment:
Generally favor `as_tibble` over `tibble` for converting objects to tibbles.  Also should probably add a comment here noting that `tibble`/`as_tibble` don't appear to copy the column data (what we want here, though it would lead to violations of `data.table` memory model in general use --- see `data.table:::as.data.frame.data.table` for what it probably wants to be happening in the general case).

Reply:
Also this is going to invisibly return the tibble-ified DT.  We want to instead `return(invisible(x))`.

Reply:
Requested improvements on tibble print formatting:

1. Can you please replace
```
# A tibble: 129,638 × 5
```
with an explanation of what is being printed?

2. Print the dimensions only once, print the column names only once.
3. Include some of the final rows, like `data.table`'s `print` method does... this will be handy to help point out latency differences between signals.

Reply:
Any pointers for how to replace the message from (1)? That seems to be a default from the tibble print method, is that customizeable?